src/token-processor/util/metadata-helpers.ts

```diff
@@ -0,0 +1,373 @@
+      continue;
+    }
+    // Process image URL with `ENV.METADATA_IMAGE_CACHE_PROCESSOR`.
+    const image =
+      metadata.image ??
+      metadata.imageUrl ??
+      metadata.image_url ??
+      defaultInsert?.metadata.image ??
+      null;
+    let cachedImage: string | null = null;
+    if (image && typeof image === 'string') {
+      const normalizedUrl = getImageUrl(image);
+      cachedImage = await processImageUrl(normalizedUrl);
+    }
+    // Localized values override defaults.
+    const metadataInsert: DbMetadataInsert = {
+      sip: 16,
+      token_id: token.id,
+      name: name.toString(),
+      description:
+        metadata.description?.toString() ?? defaultInsert?.metadata.description?.toString() ?? null,
+      image: image?.toString(),
+      cached_image: cachedImage,
+      l10n_default: raw.default,
+      l10n_locale: raw.locale ?? null,
+      l10n_uri: raw.uri,
+    };
+    // Localized attributes rewrite all default attributes. No fall back.
+    const attributes: DbMetadataAttributeInsert[] = [];
+    if (RawMetadataAttributesCType.Check(metadata.attributes)) {
+      for (const { trait_type, value, display_type } of metadata.attributes) {
+        if (trait_type && value) {
+          attributes.push({
+            trait_type: trait_type,
+            value: value,
+            display_type: display_type ?? null,
+          });
+        }
+      }
+    }
+    // Localized properties only override their default. All others have to fall back to default
+    // values.
+    const properties: DbMetadataPropertyInsert[] = defaultInsert?.properties ?? [];
+    if (RawMetadataPropertiesCType.Check(metadata.properties)) {
+      for (const [key, value] of Object.entries(metadata.properties)) {
+        if (key && value) {
+          const defaultProp = properties.find(p => p.name === key);
+          if (defaultProp) {
+            defaultProp.value = value;
+          } else {
+            properties.push({
+              name: key,
+              value: value,
+            });
+          }
+        }
+      }
+    }
+    inserts.push({
+      metadata: metadataInsert,
+      attributes: attributes,
+      properties: properties,
+    });
+    if (inserts.length === 1) {
+      defaultInsert = inserts[0];
+    }
+  }
+  return inserts;
+}
+
+/**
+ * Fetches metadata while monitoring timeout and size limits. Throws if any is reached.
+ * Taken from https://github.com/node-fetch/node-fetch/issues/1149#issuecomment-840416752
+ * @param httpUrl - URL to fetch
+ * @returns Response text
+ */
+export async function performSizeAndTimeLimitedMetadataFetch(
+  httpUrl: URL
+): Promise<string | undefined> {
+  const url = httpUrl.toString();
+  try {
+    const result = await request(url, {
+      method: 'GET',
+      throwOnError: true,
+      dispatcher:
+        // Disable during tests so we can inject a global mock agent.
+        process.env.NODE_ENV === 'test'
+          ? undefined
+          : new Agent({
```

Comment:
I think these Agent instances are intended to be re-used across requests. AFAIK the instance keeps track of re-using sockets and DNS requests? I think creating a new one for each request puts a lot more strain on system resources, like file system handles (sockets) won't be closed until nodejs garbage collection cleans up the Agent instances.

Reply:
Good catch, done